### PR TITLE
CUDA: configuration flags

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4793/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4800/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4801/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4802/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4803/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4804/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4805/ for details (Mellanox internal link).

**Comment by bureddy** [comment]
@miked-mellanox @yosefe please review

**Comment by shamisp** [comment]
👍

File: config/m4/cuda.m4 [hunk 1]
```diff
@@ -1 +1,86 @@
...
+#
+
+#
+# Check for CUDA support
+#
+cuda_happy="no"
+gdrcopy_happy="no"
+
+AC_ARG_WITH([cuda],
+           [AS_HELP_STRING([--with-cuda=(DIR)], [Enable the use of CUDA (default is no).])],
```
Comment (yosefe) [inline config/m4/cuda.m4]: why default is "no"? i think default should be "yes"/"autodetect", same as with other TLs

File: config/m4/cuda.m4 [hunk 2]
```diff
@@ -1 +1,86 @@
...
+        AC_CHECK_HEADERS([cuda.h cuda_runtime.h],
+                       [AC_CHECK_DECLS([cuPointerGetAttribute],
+                           [cuda_happy="yes"],
+                           [AC_MSG_WARN([CUDA runtime not detected. Disable.])
+                            cuda_happy="no"],
+                            [#include <cuda.h>])
+                           AS_IF([test "x$cuda_happy" == "xyes"],
+                            [AC_DEFINE([HAVE_CUDA], 1, [Enable CUDA support])
+                            AC_SUBST(CUDA_CPPFLAGS, "-I$ucx_check_cuda_dir/include ")
+                            AC_SUBST(CUDA_CFLAGS, "-I$ucx_check_cuda_dir/include ")
```
Comment (yosefe) [inline config/m4/cuda.m4]: no need to add -I to CFLAGS, just to CPPFLAGS

File: config/m4/cuda.m4 [hunk 3]
```diff
@@ -1 +1,86 @@
...
+
+AC_ARG_WITH([cuda],
+           [AS_HELP_STRING([--with-cuda=(DIR)], [Enable the use of CUDA (default is no).])],
+           [], [with_cuda=no])
+
+AS_IF([test "x$with_cuda" != "xno"],
+        [AS_IF([test ! -z "$with_cuda" -a "x$with_cuda" != "xyes"],
+            [
+            ucx_check_cuda_dir="$with_cuda"
+            ucx_check_cuda_libdir="$with_cuda/lib64 "
```
Comment (yosefe) [inline config/m4/cuda.m4]: is it ok to assume lib64?
does cuda ship pkg-config?

File: config/m4/cuda.m4 [hunk 4]
```diff
@@ -1 +1,86 @@
...
+                       [AC_MSG_WARN([CUDA not found])
+                        AC_DEFINE([HAVE_CUDA], [0], [Disable the use of CUDA])])],
+      [AC_MSG_WARN([CUDA was explicitly disabled])
+      AC_DEFINE([HAVE_CUDA], [0], [Disable the use of CUDA])]
+)
+
+
+AM_CONDITIONAL([HAVE_CUDA], [test "x$cuda_happy" != xno])
+
+AC_ARG_WITH([gdrcopy],
```
Comment (yosefe) [inline config/m4/cuda.m4]: maybe gdrcopy should be in a separate m4 file

File: config/m4/cuda.m4 [hunk 5]
```diff
@@ -1 +1,86 @@
...
+           [AS_HELP_STRING([--with-gdrcopy=(DIR)], [Enable the use of GDR_COPY (default is no).])],
```
Comment (yosefe) [inline config/m4/cuda.m4]: same here, default should be yes

File: config/m4/cuda.m4 [hunk 6]
```diff
@@ -1 +1,86 @@
...
+        [AS_IF([test ! -z "$with_cuda" -a "x$with_cuda" != "xyes"],
+            [
+            ucx_check_cuda_dir="$with_cuda"
+            ucx_check_cuda_libdir="$with_cuda/lib64 "
+            ])
+        AS_IF([test ! -z "$with_cuda_libdir" -a "x$with_cuda_libdir" != "xyes"],
+            [ucx_check_cuda_libdir="$with_nccl_libdir"])
+
+        AC_CHECK_HEADERS([cuda.h cuda_runtime.h],
+                       [AC_CHECK_DECLS([cuPointerGetAttribute],
```
Comment (yosefe) [inline config/m4/cuda.m4]: maybe use AC_CHECK_FUNC which also checks link, and not just compile?

File: contrib/test_jenkins.sh [hunk 7]
```diff
@@ -239,6 +239,36 @@ build_debug() {
...
+#
+
+build_cuda() {
+    echo 1..1 > build_cuda.tap
+    if module_load dev/cuda
+    then
+        if module_load dev/gdrcopy
+        then
+            echo "==== Build with enable cuda ===="
+            ../contrib/configure-devel --prefix=$ucx_inst --with-cuda --with-gdrcopy
```
Comment (yosefe) [inline contrib/test_jenkins.sh]: do we need to test that cuda was actually built?

**Comment by yosefe** [comment]
bot:bgate:retest

File: config/m4/cuda.m4 (hunk 1 above)
Comment (bureddy) [reply config/m4/cuda.m4]: done

File: config/m4/cuda.m4 (hunk 3 above)
Comment (bureddy) [reply config/m4/cuda.m4]: checked for 64 suffix

File: config/m4/cuda.m4 (hunk 6 above)
Comment (bureddy) [reply config/m4/cuda.m4]: changed to AC_CHECK_LIB

File: contrib/test_jenkins.sh (hunk 7 above)
Comment (bureddy) [reply contrib/test_jenkins.sh]: configure fails if not able to build cuda, do we need more than this?

**Comment by bureddy** [comment]
@yosefe please review it again.

**Comment by bureddy** [comment]
@yosefe sorry I amended the PR instead of new commit

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4817/ for details (Mellanox internal link).

**Pull request merged**